core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1178,39 +1335,86 @@ private void listStatusInternal(ListStatusContext context, RpcContext rpcContext
           ? DescendantType.ALL : DescendantType.NONE;
-      // This is to generate a parsed child path components to be passed to lockChildPath
-      String [] childComponentsHint = null;
-      try (CloseableIterator<? extends Inode> it = mInodeStore.getChildren(inode.asDirectory())) {
-        while (it.hasNext()) {
-          Inode child = it.next();
+      try (CloseableIterator<? extends Inode> childrenIterator = getChildrenIterator(
+          inode, partialPath, prefixComponents, depth, context)) {
+        // This is to generate a parsed child path components to be passed to lockChildPath
+        String[] childComponentsHint = null;
+        while (childrenIterator.hasNext()) {
+          if (context.donePartialListing()) {
+            return;
+          }
+          String childName = childrenIterator.next().getName();
           if (childComponentsHint == null) {
             String[] parentComponents = PathUtils.getPathComponents(
                 currInodePath.getUri().getPath());
             childComponentsHint = new String[parentComponents.length + 1];
             System.arraycopy(parentComponents, 0, childComponentsHint, 0, parentComponents.length);
           }
           // TODO(david): Make extending InodePath more efficient
-          childComponentsHint[childComponentsHint.length - 1] = child.getName();
+          childComponentsHint[childComponentsHint.length - 1] = childName;
 
           try (LockedInodePath childInodePath =
-                   currInodePath.lockChild(child, LockPattern.READ, childComponentsHint)) {
-            listStatusInternal(context, rpcContext, childInodePath,
-                auditContext, nextDescendantType,
-                resultStream, depth + 1, counter);
+                   currInodePath.lockChildByName(
+                       childName, LockPattern.READ, childComponentsHint)) {
+            listStatusInternal(context, rpcContext, childInodePath, auditContext,
+                nextDescendantType, resultStream, depth + 1, counter,
+                partialPath, prefixComponents);
           } catch (InvalidPathException | FileDoesNotExistException e) {
-            LOG.debug("Path \"{}\" is invalid, has been ignored.",
-                PathUtils.concatPath("/", childComponentsHint));
+            if (LOG.isDebugEnabled()) {
+              LOG.debug("Path \"{}\" is invalid, has been ignored.",
+                  PathUtils.concatPath("/", (Object) childComponentsHint));
+            }
           }
+          // ensure that the listing of any new subdirectories starts from the
+          // beginning of the children list
+          partialPath = null;
         }
       }
     }
-    // Listing a directory should not emit item for the directory itself.
-    if (depth != 0 || inode.isFile()) {
-      resultStream.submit(getFileInfoInternal(currInodePath, counter));
+  }
+
+  private CloseableIterator<? extends Inode> getChildrenIterator(
```

Comment:
One thought I have is we can maybe try not to make DefaultFileSystemMaster any longer, by trying to move the underlying impl somewhere else in `PartialListingIterator` for example. This file is formidably long now.